### PR TITLE
PERL-747: Extended JSON

### PR DESCRIPTION
**Pull request by phaylon**
Implements Extended JSON and adds the new BSON corpus tests.

### CODE REVIEW

File: lib/BSON.pm [hunk 1]
```diff
@@ -437,6 +439,261 @@ sub inflate_extjson {
     return $hash;
 }
 
+=method perl_to_extjson
+
+    use JSON::MaybeXS;
+    my $ext = $bson->perl_to_extjson($data);
```
Comment (xdg) [inline lib/BSON.pm]: Ought to be: `BSON->perl_to_extjson($data, $options)`

File: lib/BSON.pm [hunk 2]
```diff
@@ -437,6 +439,261 @@ sub inflate_extjson {
...
+    my $json = encode_json($ext, \%options);
```
Comment (xdg) [inline lib/BSON.pm]: No `\%options` here, I think.

File: lib/BSON.pm [hunk 3]
```diff
@@ -437,6 +439,261 @@ sub inflate_extjson {
...
+            return $data;
+        }
+    }
+
+    if (boolean::isBoolean($data)) {
+        return $data;
+    }
+
+    if (ref $data eq 'HASH') {
+        if (exists $data->{'$type'}) {
```
Comment (xdg) [inline lib/BSON.pm]: I'm not sure this is needed.  The special handling for `$type` is needed when parsing ExtJSON but shouldn't need any special work for encoding.  (Unless this is handling some recursive situation from encoding Bytes/Binary)

File: lib/BSON.pm [hunk 4]
```diff
@@ -437,6 +439,261 @@ sub inflate_extjson {
...
+specific data structure. Note that you have to decode the JSON string
+manually beforehand.
+
+=cut
+
+sub extjson_to_perl {
+    my ($class, $data) = @_;
+
+    if (ref $data eq 'HASH') {
+        my $hash = $data;
```
Comment (xdg) [inline lib/BSON.pm]: I don't understand what this is accomplishing. Just renaming?

File: lib/BSON.pm [hunk 5]
```diff
@@ -437,6 +439,261 @@ sub inflate_extjson {
...
+
+=method extjson_to_perl
+
+    use JSON::MaybeXS;
+    my $ext = decode_json($json);
+    my $data = $bson->extjson_to_perl($ext);
+
+Takes an Extended JSON data structure and inflates it into a Perl
+specific data structure. Note that you have to decode the JSON string
+manually beforehand.
```
Comment (xdg) [inline lib/BSON.pm]: Pls document the conditions under which this will wrap numbers for type preservation.  It looks like it will wrap anything expressed "canonically" as a hashref and won't wrap anything expressed as a scalar. Is that right?

File: lib/BSON/DBRef.pm [hunk 6]
```diff
@@ -130,12 +130,23 @@ sub TO_JSON {
...
-        return {
-            '$ref' => $self->ref,
-            '$id'  => $self->id,
-            ( defined($self->db) ? ( '$db' => $self->db ) : () ),
-            %{ $self->extra },
-        };
+        my $id = $self->id;
+        $id = ref($id)
+            ? $id->TO_JSON
+            : { '$numberInt' => "$id" };
```
Comment (xdg) [inline lib/BSON/DBRef.pm]: I think this needs to do "looks_like_number", etc. and set this to be $numberInt, $numberLong, $numberDouble or else just put it in as an ordinary string.

File: lib/BSON/Double.pm [hunk 7]
```diff
@@ -50,9 +50,23 @@ my $illegal = $^O eq 'MSWin32' && $] lt "5.022" ? qr/^$win32_specials/ : qr/^$un
 
 sub TO_JSON {
     my $copy = "$_[0]->{value}"; # avoid changing value to PVNV
-    return $_[0]->{value}/1.0 unless $copy =~ $illegal;
 
-    croak( "The value '$copy' is illegal in JSON" );
+    return { '$numberDouble' => 'Infinity' }
+        if $copy eq 'Inf';
```
Comment (xdg) [inline lib/BSON/Double.pm]: This isn't portable here and below.  You'll get text that matches `$win32_specials` on windows and it won't compare equal here.

File: t/lib/CorpusTest.pm [hunk 8]
```diff
@@ -64,56 +104,150 @@ sub _validity_tests {
...
+            my $has_canonical_bson = defined $canonical_bson;
+            my $has_converted_bson = defined $converted_bson;
+            my $has_degenerate_bson = defined $degenerate_bson;
+
+            my $has_canonical_json = defined $canonical_json;
+            my $has_converted_json = defined $converted_json;
+            my $has_relaxed_json = defined $relaxed_json;
+            my $has_degenerate_json = defined $degenerate_json;
+
+            if (!$deprecated and $has_canonical_bson and $has_canonical_json) {
```
Comment (xdg) [inline t/lib/CorpusTest.pm]: I find it hard to follow all the conditionals and map them back to the spec assertions.  Could you please refactor them at least into two functions: one for deprecated and one for non-deprecated.

Then I think some of the conditionals are unnecessary as they are certain to exist (e.g. canonical_bson and canonical_json).

File: t/lib/CorpusTest.pm [hunk 9]
```diff
@@ -4,11 +4,13 @@ use warnings;
 use Test::More 0.96;
 use Test::Deep qw/!blessed/;
 
+use JSON::PP ();
+use JSON::XS ();
```
[the remark this answers is not in the document]
Comment (phaylon) [reply t/lib/CorpusTest.pm]: I've actually now switched it to `Cpanel::JSON::XS`. The `XS`/`PP` usage is because we require `PP` for the ability to created an `IxHash`, but we need `XS` for proper svtype detection. The leftover usage of `MaybeXS` was from earlier, and I'm now using the `XS` variant explicitly.

File: t/lib/CorpusTest.pm [hunk 10]
```diff
@@ -64,56 +104,150 @@ sub _validity_tests {
...
-
-        my $skip_extjson = !(defined($E) && _extjson_ok($bson_type, $E));
-
-        $B = pack( "H*", $B );
-        $cB = pack( "H*", $cB );
-
-        $E = _normalize( $E, "$desc: normalizing E"  );
-        $cE = _normalize( $cE, "$desc: normalizing cE"  );
-
-        _bson_to_bson( $codec, $B, $cB, "$desc: B->cB" );
```
[the remark this answers is not in the document]
Comment (phaylon) [reply t/lib/CorpusTest.pm]: Since the former versions of these used `bson` and `canonical_bson` I'm unsure how they would map to the new data fields. I've added a BSON roundtrip test for canonical/converted now, but I'm unsure if this is sufficient.

File: t/lib/CorpusTest.pm (hunk 9 above)
[the remark this answers is not in the document]
Comment (xdg) [reply t/lib/CorpusTest.pm]: Can you talk this over with @shadowcat-mst please?  Keeping BSON pure-Perl capable -- even in testing -- is pretty important; part of its reason for being, even.  If there's no way to avoid XS for testing, then it needs to be conditional.

File: t/lib/CorpusTest.pm [hunk 11]
```diff
@@ -148,109 +129,144 @@ sub _validity_tests {
...
+    my ($canonical_json, $degenerate_json, $relaxed_json)
+        = @{$spec}{qw( canonical_json degenerate_json relaxed_json )};
+
+    my $lossy = $spec->{lossy};
+
+    _bson_to_bson(
+        $codec,
+        $canonical_bson,
+        defined($converted_bson) ? $converted_bson : $canonical_bson,
+        'cB/B -> cB',
```
Comment (xdg) [inline t/lib/CorpusTest.pm]: Nit: `cB -> cB`

File: t/lib/CorpusTest.pm [hunk 12]
```diff
@@ -148,109 +129,144 @@ sub _validity_tests {
...
-            if (!$deprecated and $has_degenerate_bson and $has_canonical_json) {
-                _bson_to_extjson(
-                    $codec,
-                    $degenerate_bson,
-                    $canonical_json,
-                    'dB -> cEJ',
-                );
-            }
+# deprecated and non-deprecated
+sub _validity_tests_all {
```
Comment (xdg) [inline t/lib/CorpusTest.pm]: I still find these very hard to reconcile against the bson-corpus spec.  Could you please try reorganizing to more directly follow the structure in the spec?  (e.g. "For cB input...", "For cEJ input..."?

**Comment by xdg** [comment]
Squashed and merged as c863549238ad7d3f009e925090ac397e2efe2972
